cluster/gce/windows/common.psm1

```diff
@@ -146,6 +146,320 @@ function MustDownload-File {
+# On failure, throws an exception.
+function Close-Handle {
+  param (
+    [parameter(Mandatory=$true)] [System.IntPtr]$Handle
+  )
+  $ret = $Kernel32::CloseHandle($Handle)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if (-not $ret) {
+    throw "Failed to close open file handle ${Handle}, system error code: ${err}"
+  }
+}
+
+# Open-File tries to open the file at the specified path with ReadWrite access mode and ReadWrite file share mode.
+# On success, returns an open file handle.
+# On failure, throws an exception.
+function Open-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+
+  $lpFileName = $Path
+  $dwDesiredAccess = [System.IO.FileAccess]::ReadWrite
+  $dwShareMode = [System.IO.FileShare]::ReadWrite # Fortunately golang also passes these same flags when it creates the log files, so we can open it concurrently.
+  $lpSecurityAttributes = [System.IntPtr]::Zero
+  $dwCreationDisposition = [System.IO.FileMode]::Open
+  $dwFlagsAndAttributes = [System.IO.FileAttributes]::Normal
+  $hTemplateFile = [System.IntPtr]::Zero
+
+  $handle = $Kernel32::CreateFileW($lpFileName, $dwDesiredAccess, $dwShareMode, $lpSecurityAttributes, $dwCreationDisposition, $dwFlagsAndAttributes, $hTemplateFile)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if ($handle -eq -1) {
+    throw "Failed to open file ${Path}, system error code: ${err}"
+  }
+
+  return $handle
+}
+
+# Truncate-File truncates the file in-place by opening it, moving the file pointer to the beginning,
+# and setting the end of file to the file pointer's location.
+# On failure, throws an exception.
+# The file must have been originally created with FILE_SHARE_WRITE for this to be possible.
+# Fortunately Go creates files with FILE_SHARE_READ|FILE_SHARE_WRITE by for all os.Open calls,
+# so our log writers should be doing the right thing.
+function Truncate-File {
+  param (
+    [parameter(Mandatory=$true)] [string]$Path
+  )
+  $INVALID_SET_FILE_POINTER = 0xffffffff
+  $NO_ERROR = 0
+  $FILE_BEGIN = 0
+
+  $handle = Open-File -Path $Path
+
+  # https://docs.microsoft.com/en-us/windows/desktop/api/fileapi/nf-fileapi-setfilepointer
+  # Docs: Because INVALID_SET_FILE_POINTER is a valid value for the low-order DWORD of the new file pointer,
+  # you must check both the return value of the function and the error code returned by GetLastError to
+  # determine whether or not an error has occurred. If an error has occurred, the return value of SetFilePointer
+  # is INVALID_SET_FILE_POINTER and GetLastError returns a value other than NO_ERROR.
+  $ret = $Kernel32::SetFilePointer($handle, 0, [System.IntPtr]::Zero, $FILE_BEGIN)
+  $err = [System.Runtime.InteropServices.Marshal]::GetLastWin32Error()
+  if ($ret -eq $INVALID_SET_FILE_POINTER -and $err -ne $NO_ERROR) {
+    Close-Handle -Handle $handle
+    throw "Failed to set file pointer for handle ${handle}, system error code: ${err}"
```

Comment:
How are the exceptions handled in the script? Will the script terminate and does it log anywhere?

Reply:
An exception (terminating error) should terminate the rotation script. The behavior in the startup scripts has been to terminate and log the exception to the serial console (assuming it was stderr). There are also non-terminating errors that only crash the script if `$ErrorActionPreference = 'Stop'` (it's `Continue` by default, and in the rotation script, though we set it to `Stop` in `configure.ps1`). Non-terminating errors are things like `ObjectNotFound`.

I'm not sure where logs go, if any, from scheduled tasks. If the task crashes due to the exception, there *may* be a message in the Windows event logs.

One thing I haven't done but could: Catch exceptions from Rotate-File in Rotate-Files and log them, so issues rotating one file won't block others. If I do this, however, it may reduce our ability to detect errors, since the rotation script won't crash and there may be no stderr/stdout logs anywhere.

Reply:
Decided to set Stop in the rotation script and catch those exceptions. I'd rather the script be more robust since error detection for scheduled tasks looks difficult.

Reply:
To be clear, failed to rotate one file won't affect other files, right?

Reply:
It shouldn't now that we catch exceptions from Rotate-File in Rotate-Files.